### PR TITLE
Browser test

### PR DESCRIPTION
**Pull request by cancerberoSgx**
* WIP - created this pull request so we can discuss this (big) changes.
 * move magickApi.js to src/magickApi.ts
 * ./spec contain jasmine unit tests that run in the browser thanks to gulp-jasmine-browser. 
   * Currently running with chrome (puppeteer) but it also support gecko (slimerjs) and webkit (phantomjs)
   * `npm run test-browser` will run browser headless in terminal
   * `npm run test-browser-server`will run tests in a server so yor can open chrome at http://localhost:8888 to debug them
   * TODO: use watch/browsersync to agile development
* ./src/util/file.ts contains utilities to easy MagickInputFile creation and transformation between input files and output files, blobs, arrays, strings, etc
* ./src/util/image.ts contains utilities related to images. Right now has compare() and extractInfo() that help me in the tests
 * please ignore src/imageHome and src/execute -  those features (and other) will probably have to be added in another branch / project

### CODE REVIEW

**Comment by cancerberoSgx** [comment]
Hey @KnicKnic I'm using branches your own repo instead of my forked repo using the permissions you give me. Created this pull request because these changes are big and I need you to review them first and give me permission to merge it to master.

I need this codebase with a src folder and BDD specs that run in the browser  before adding new features to the project. This spec uses a syntax with describe()/it()/expect() that major testing libraries support : mocha, jasmine, jest, etc. Run in the browser thanks to gulp-jasmine-browser currently in chrome headless (puppeteer) and supposedly they run fine in travis CI and others. 

Don't know where the current tests in test/rotate fit here - perhaps I can port them as jasmine spec and run them in the browser so we have all tests in the same infrastructure / source folder ?

File: Readme.md [hunk 1]
```diff
@@ -109,7 +110,15 @@ docker run --rm -it --workdir /code -v "$PWD":/code wasm-imagemagick-build-tools
 
 #windows cmd
 #docker run --rm -it --workdir /code -v %CD%:/code wasm-imagemagick-build-tools bash ./build.sh
-
 ```
 
-Produces `magick.js` & `magick.wasm` in the current folder. Along with `magickApi.js` that is all the files needed on the webserver.
```
Comment (KnicKnic) [inline Readme.md]: Please add how & where magickApi.js is produced. Is it in /src after running npm run build?
also can you update the publish code to include this
https://github.com/KnicKnic/WASM-ImageMagick/blob/ccecb8bfb9e531f0ee3c3932e97c6a94bdba14e0/azure-pipelines.yml#L25-L26
if you look two blocks up you in the azure pipelines file you can see how to do multiple statements.

Comment (KnicKnic) [reply Readme.md]: Also I think your tests will fail till you use npm install here
https://github.com/KnicKnic/WASM-ImageMagick/blob/ccecb8bfb9e531f0ee3c3932e97c6a94bdba14e0/azure-pipelines.yml#L14-L15

Comment (cancerberoSgx) [reply Readme.md]: Didn't support to azure-pipelines CI yet - npm test command is the same as before.  - first wanted to see if this is acceptable to you and review - then will try to make the CI to work. thanks (still pushing changes)

Comment (cancerberoSgx) [reply Readme.md]: > Please add how & where magickApi.js is produced. Is it in /src after running npm run build?
> also can you update the publish code to include this

you are right - magickApi.js that can be loaded in the browser directly is no longer supported. I must compile src/magickApi.ts separately / with different configuration so it's generated.

Comment (KnicKnic) [reply Readme.md]: > Didn't support to azure-pipelines CI yet - npm test command is the same as before. - first wanted to see if this is acceptable to you and review - then will try to make the CI to work. thanks (still pushing changes)

BTW I am very happy with what you have done. I am okay with the set of changes in general. If you do not want to learn another CI I can do that work.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -7,21 +7,22 @@ pool:
   vmImage: 'Ubuntu 16.04'
 
 steps:
-- script: |
-    sudo npm install -g typescript
```
Comment (cancerberoSgx) [inline azure-pipelines.yml]: no need to install anything globally - commands to run are npm install && npm test

File: azure-pipelines.yml [hunk 3]
```diff
@@ -7,21 +7,22 @@ pool:
...
 #    cd tests/rotate
 #    docker run -v "$PWD":/code --rm --workdir /code node node  /code/node.js
 #  displayName: test wasm with nodejs
 
-- script: npm run test 
+- script: npm test
   displayName: run tests
   
+
+- script: npm run test-browser
```
Comment (cancerberoSgx) [inline azure-pipelines.yml]: working ! : https://dev.azure.com/oneeyedelf1/wasm-imagemagick/_build/results?buildId=78&view=logs

maybe we want that npm test to run both things so we have a single `npm test` command here ?

**Comment by cancerberoSgx** [comment]
an update: 

 * azure-pipelines now
   * uses latest node and npm (install globally  using nave before `npm instal`)
   * runs browser tests and node tests

* high level execute() function that:
   * supports array commands syntax ['convert', 'foo.png','bar.png'] and CLI like commands i.e: 'convert foo.png bar.png'
   * supports running several commands serially reusing output images as input images in following commands (see spec/executeSpec.ts)
* spec/formatSpec.ts checks that all supported formats work
* more utilities: cli and html 

## TODO
 
 * documen high level API execute in readme
 * generate magickApi.js to support direct import from html

**Comment by cancerberoSgx** [comment]
@KnicKnic  I think this PR is ready merge. I don't want to keep adding features here without merging this changes first. What do you think ?

**Comment by KnicKnic** [comment]
> @KnicKnic I think this PR is ready merge. I don't want to keep adding features here without merging this changes first. What do you think ?

Looks good, only one small nit I would like fixed. Can you ensure that magickApi.js still gets copied to the artifacts. It was getting coppied from
https://github.com/KnicKnic/WASM-ImageMagick/blob/be436af4380a63b8e110da4631eac89addea387f/azure-pipelines.yml#L39-L40
but since magickApi.js is no longer in the root folder it is no longer copied in this stage.

Thanks,
Knic

**Comment by cancerberoSgx** [comment]
> but since magickApi.js is no longer in the root folder it is no longer copied in this stage.

Done, apiMagick.js is generated in the root folder now, thanks. Waiting for your final permission to merge this (unless you want to do it yourself). Thanks

**Comment by cancerberoSgx** [comment]
Merged, the only thing missing would be upgrading package.json version number and

**Pull request merged**